ipfs-retriever/bin/ipfs-retriever.js

```diff
@@ -0,0 +1,323 @@
+      })
+
+      if (!responseBody) {
+        // The upstream response does not have any readable body
+        // There is no need to measure response body size, we can
+        // return the original response object.
+        ctx.waitUntil(
+          logRetrievalResult(env, {
+            cacheMiss,
+            responseStatus: originResponse.status,
+            egressBytes: 0,
+            requestCountryCode,
+            timestamp: requestTimestamp,
+            dataSetId,
+          }),
+        )
+        const response = new Response(originResponse.body, originResponse)
+        setContentSecurityPolicy(response)
+        response.headers.set('X-Data-Set-ID', dataSetId)
+        response.headers.set(
+          'Cache-Control',
+          `public, max-age=${env.CLIENT_CACHE_TTL}`,
+        )
+        return response
+      }
+
+      // Stream and count bytes
+      // We create two identical streams, one for the egress measurement and the other for returning the response as soon as possible
+      const [returnedStream, egressMeasurementStream] = responseBody.tee()
+      const reader = egressMeasurementStream.getReader()
+      const firstByteAt = performance.now()
+
+      ctx.waitUntil(
+        (async () => {
+          const egressBytes = await measureStreamedEgress(reader)
+          const lastByteFetchedAt = performance.now()
+
+          await logRetrievalResult(env, {
+            cacheMiss,
+            responseStatus: originResponse.status,
+            egressBytes,
+            requestCountryCode,
+            timestamp: requestTimestamp,
+            performanceStats: {
+              fetchTtfb: firstByteAt - fetchStartedAt,
+              fetchTtlb: lastByteFetchedAt - fetchStartedAt,
+              workerTtfb: firstByteAt - workerStartedAt,
+            },
+            dataSetId,
+          })
+
+          await updateDataSetStats(env, { dataSetId, egressBytes })
+        })(),
+      )
+
+      // Return immediately, proxying the transformed response
+      const response = new Response(returnedStream, {
+        status: originResponse.status,
+        statusText: originResponse.statusText,
+        headers: originResponse.headers,
+      })
+      setContentSecurityPolicy(response)
+      response.headers.set('X-Data-Set-ID', dataSetId)
+      response.headers.set(
+        'Cache-Control',
+        `public, max-age=${env.CLIENT_CACHE_TTL}`,
+      )
+
+      // FIXME: move this logic into processIpfsResponse function
+      // When converting from CAR to RAW, set content-disposition to inline
+      // so browsers display the content instead of downloading it.
+      if (ipfsFormat !== 'car') {
+        response.headers.set('content-disposition', 'inline')
+        // Also remove the content-type header, remove x-content-type-options,
+        // and let the browser to sniff the content type.
+        response.headers.delete('content-type')
+        response.headers.delete('x-content-type-options')
+      }
```

Comment:
After we move this block to `processIpfsResponse()`, the code post-processing the response should be pretty much identical in both piece-retriever and ipfs-retriever.